File: ckgame-java/src/main/java/com/ckgame/military/WarManager.java
package com.ckgame.military;

import com.ckgame.core.Constants;
import com.ckgame.core.calendar.GameDate;
import com.ckgame.politics.CasusBelli;
import com.ckgame.politics.Diplomacy;
import com.ckgame.world.World;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.ToDoubleFunction;

/**
 * 管理所有战争与军队。
 */
public final class WarManager {
    public final Map<Integer, War> wars = new HashMap<>();
    public final Map<Integer, Army> armies = new HashMap<>();
    public int nextWar = 1;
    public int nextArmy = 1;

    private final World world;
    private final Diplomacy diplomacy;

    public WarManager() {
        this(null, null);
    }

    public WarManager(World world) {
        this(world, null);
    }

    public WarManager(World world, Diplomacy diplomacy) {
        this.world = world;
        this.diplomacy = diplomacy;
    }

    public World world() {
        return world;
    }

    public Diplomacy diplomacy() {
        return diplomacy;
    }

    /** 宣战并创建战争。 */
    public int declareWar(CasusBelli cb, int attacker, int defender,
                          GameDate date, String name, int targetTitle) {
        int wid = nextWar++;
        War war = new War(wid, name, cb, attacker, defender, date);
        war.targetTitle = targetTitle;
        war.participants.add(new WarParticipant(attacker, true, date));
        war.participants.add(new WarParticipant(defender, false, date));
        wars.put(wid, war);
        return wid;
    }

    public int declareWar(CasusBelli cb, int attacker, int defender,
                          GameDate date, String name) {
        return declareWar(cb, attacker, defender, date, name, Constants.NONE_ID);
    }

    /** 按 id 获取战争。 */
    public War war(int wid) {
        return wars.get(wid);
    }

    /** 所有进行中的战争。 */
    public List<War> activeWars() {
        List<War> out = new ArrayList<>();
        for (War w : wars.values()) {
            if (w.active) {
                out.add(w);
            }
        }
        return out;
    }

    /** 结束战争并设置结果。 */
    public void endWar(int wid, WarResult result) {
        War w = wars.get(wid);
        if (w == null) {
            return;
        }
        w.active = false;
        w.result = result;
        // 战争结束时处理领土变更（Python 版此处为占位 pass）
        if (result == WarResult.ATTACKER_VICTORY) {
            // 攻击者获胜，夺取目标省份
        } else if (result == WarResult.DEFENDER_VICTORY) {
            // 防御者获胜，击退攻击者
        }
    }

    /** 征召一支新军队。 */
    public int raiseArmy(int owner, int location, int levies, String name) {
        int aid = nextArmy++;
        String armyName = (name == null || name.isBlank()) ? "军团#" + aid : name;
        Army army = new Army(aid, owner, armyName, location, owner);
        army.addMen(UnitType.LEVIES, levies);
        armies.put(aid, army);
        return aid;
    }

    public int raiseArmy(int owner, int location, int levies) {
        return raiseArmy(owner, location, levies, "");
    }

    /** 按 id 获取军队。 */
    public Army army(int aid) {
        return armies.get(aid);
    }

    /** 某角色所有活跃军队的 id。 */
    public List<Integer> armiesOf(int owner) {
        List<Integer> out = new ArrayList<>();
        for (Army a : armies.values()) {
            if (a.owner == owner && a.isActive()) {
                out.add(a.id);
            }
        }
        return out;
    }

    /** 某角色所有活跃军队的总兵力。 */
    public int totalMenOf(int owner) {
        int total = 0;
        for (Army a : armies.values()) {
            if (a.owner == owner && a.isActive()) {
                total += a.totalMen();
            }
        }
        return total;
    }

    /** 推进所有处于行军状态的军队。 */
    public void tickMovement(ToDoubleFunction<Army> moveChanceOf) {
        for (Army army : armies.values()) {
            if (army.status == ArmyStatus.MOVING) {
                double chance = moveChanceOf == null ? 1.0 : moveChanceOf.applyAsDouble(army);
                army.advanceMove(chance);
            }
        }
    }

    public void tickMovement() {
        tickMovement(null);
    }

    /** 解散空军队。 */
    public void disbandEmpty() {
        for (Army army : armies.values()) {
            if (army.totalMen() <= 0) {
                army.status = ArmyStatus.DISBANDED;
            }
        }
    }

    /** 为指定军队设置行军路径。 */
    public void move(int armyId, List<Integer> path) {
        Army army = armies.get(armyId);
        if (army != null) {
            army.setPath(path);
        }
    }

    /** 解散指定军队。 */
    public void disband(int armyId) {
        Army army = armies.get(armyId);
        if (army != null) {
            army.status = ArmyStatus.DISBANDED;
        }
    }

    /** 一日推进：移动与清理空军队。 */
    public void tick() {
        tickMovement();
        disbandEmpty();
    }
}

File: ckgame-ts/src/military/WarManager.ts
import { NONE_ID } from '../core/Constants.js';
import { GameDate } from '../core/calendar/GameDate.js';
import { CasusBelli } from '../politics/CasusBelli.js';
import type { Diplomacy } from '../politics/Diplomacy.js';
import type { World } from '../world/World.js';
import { Army } from './Army.js';
import { ArmyStatus } from './ArmyStatus.js';
import { UnitType } from './UnitType.js';
import { War } from './War.js';
import { WarParticipant } from './WarParticipant.js';
import { WarResult } from './WarResult.js';

/**
 * War manager: manages all wars and armies.
 */
export class WarManager {
  public wars: Map<number, War> = new Map();
  public armies: Map<number, Army> = new Map();
  public nextWar: number = 1;
  public nextArmy: number = 1;

  private world: World | null;
  private diplomacy: Diplomacy | null;

  constructor(world?: World | null, diplomacy?: Diplomacy | null) {
    this.world = world ?? null;
    this.diplomacy = diplomacy ?? null;
  }

  getWorld(): World | null {
    return this.world;
  }

  getDiplomacy(): Diplomacy | null {
    return this.diplomacy;
  }

  /** Declare war and create a war entry. */
  declareWar(cb: CasusBelli, attacker: number, defender: number, date: GameDate, name: string, targetTitle?: number): number;
  declareWar(cb: CasusBelli, attacker: number, defender: number, date: GameDate, name: string): number;
  declareWar(cb: CasusBelli, attacker: number, defender: number, date: GameDate, name: string, targetTitle: number = NONE_ID): number {
    const wid = this.nextWar++;
    const war = new War(wid, name, cb, attacker, defender, date);
    war.targetTitle = targetTitle;
    war.participants.push(new WarParticipant(attacker, true, date));
    war.participants.push(new WarParticipant(defender, false, date));
    this.wars.set(wid, war);
    return wid;
  }

  /** Get a war by id. */
  war(wid: number): War | undefined {
    return this.wars.get(wid);
  }

  /** All active wars. */
  activeWars(): War[] {
    const out: War[] = [];
    for (const w of this.wars.values()) {
      if (w.active) out.push(w);
    }
    return out;
  }

  /** End a war with the given result. */
  endWar(wid: number, result: WarResult): void {
    const w = this.wars.get(wid);
    if (!w) return;
    w.active = false;
    w.result = result;
  }

  /** Raise a new army. */
  raiseArmy(owner: number, location: number, levies: number, name?: string): number;
  raiseArmy(owner: number, location: number, levies: number): number;
  raiseArmy(owner: number, location: number, levies: number, name: string = ''): number {
    const aid = this.nextArmy++;
    const armyName = (!name || name.trim() === '') ? `军团#${aid}` : name;
    const army = new Army(aid, owner, armyName, location, owner);
    army.addMen(UnitType.LEVIES, levies);
    this.armies.set(aid, army);
    return aid;
  }

  /** Get an army by id. */
  army(aid: number): Army | undefined {
    return this.armies.get(aid);
  }

  /** All active army ids for a given owner. */
  armiesOf(owner: number): number[] {
    const out: number[] = [];
    for (const a of this.armies.values()) {
      if (a.owner === owner && a.isActive()) out.push(a.id);
    }
    return out;
  }

  /** Total men across all active armies for a given owner. */
  totalMenOf(owner: number): number {
    let total = 0;
    for (const a of this.armies.values()) {
      if (a.owner === owner && a.isActive()) total += a.totalMen();
    }
    return total;
  }

  /** Tick movement for all moving armies. */
  tickMovement(moveChanceOf?: (army: Army) => number): void {
    for (const army of this.armies.values()) {
      if (army.status === ArmyStatus.MOVING) {
        const chance = moveChanceOf ? moveChanceOf(army) : 1.0;
        army.advanceMove(chance);
      }
    }
  }

  /** Disband empty armies. */
  disbandEmpty(): void {
    for (const army of this.armies.values()) {
      if (army.totalMen() <= 0) {
        army.status = ArmyStatus.DISBANDED;
      }
    }
  }

  /** Set path for an army. */
  move(armyId: number, path: number[]): void {
    const army = this.armies.get(armyId);
    if (army) army.setPath(path);
  }

  /** Disband an army. */
  disband(armyId: number): void {
    const army = this.armies.get(armyId);
    if (army) army.status = ArmyStatus.DISBANDED;
  }

  /** Daily tick: movement + cleanup. */
  tick(): void {
    this.tickMovement();
    this.disbandEmpty();
  }
}
